File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int ans=0,start=0,end=1;
        while(end<prices.size()){
            if(prices[start]>=prices[end]){
                start=end;
            }else if(prices[start]<prices[end]){
                ans=max(ans,prices[end]-prices[start]);
            }
            end++;
        }
        return ans;
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        // technique to buy on first day and then che
        int n = prices.length;
        int sp = prices[0];
        int maxProfit = 0;
        for(int i=1;i<n;i++){
            if(prices[i]<sp){
                sp = prices[i];
            }else if(prices[i]-sp>maxProfit){
                maxProfit = Math.max(maxProfit, prices[i]-sp);
            }
        }
        return maxProfit;
    }
}